File: Java/src/main/java/Solution501_600/Solution501.java
package Solution501_600;

import tool.TreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * @author windows
 */
public class Solution501 {
    HashMap<Integer, Integer> map = new HashMap<>();
    public int[] findMode(TreeNode root) {
        process(root);
        int max = 0;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            max = Math.max(max, entry.getValue());
        }
        ArrayList<Integer> list = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() == max) {
                list.add(entry.getKey());
            }
        }
        return Arrays.stream(list.toArray(new Integer[list.size()])).mapToInt(Integer::intValue).toArray();
    }

    private void process(TreeNode root){
        if (root == null) {
            return;
        }
        map.put(root.val, map.getOrDefault(root.val, 0) + 1);
        process(root.left);
        process(root.right);
    }


    public static void main(String[] args) {
        Solution501 solution501 = new Solution501();
        System.out.println(solution501.findMode(new TreeNode(0)));
    }
}

File: C++/src/Solution501_600/Solution501.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution501.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;


//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};



vector<int> findMode(TreeNode* root) {
  vector<int> res;
  function<void(TreeNode*, int, int)> dfs = [&](TreeNode* root, int num, int count){
    if (!root) return;





  };





}